### PR TITLE
Add dotnet PE support

### PR DESCRIPTION
**Pull request by wingdeans**
<details>
<summary><strong>Checklist</strong></summary>

- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)
</details>

## Description

Partially addresses [CIL/MSIL (.NET) bytecode support](https://github.com/rizinorg/ideas/issues/6).

## Tasks

- [X] Parse CLR headers (partially implemented from previous attempt: [.NET support in R2](https://github.com/radareorg/radare2/pull/6771))
- [x] Parse CLR metatables
- [x] Disassembler rewrite
- [x] Integrate into Rizin
  - [x] Access methods
  - [x] Access classes

**Future:**
- [x] Analysis plugin

The primary goal is to handle method headers

# Resources
- [the dotnet format](http://www.ntcore.com/files/dotnetformat.htm)
- [yara parser implementation](https://github.com/VirusTotal/yara/blob/d13bc3475518e93a472c5d9a24106dba77ea2cce/libyara/modules/dotnet/dotnet.c)
- [dotnet opcodes](https://github.com/dotnet/runtime/blob/7bac4e8fa611eb02ac1bd27b97a8570cbf321437/src/coreclr/inc/opcode.def)

# Test plan

Test on cil executables in [https://github.com/rizinorg/rizin-testbins](testbins) using `monodis` as reference
Test parsed output compared to `yara -D`

### CODE REVIEW

**Comment by wargio** [comment]
please define the spdx headers: https://spdx.org/licenses/

File: librz/bin/format/pe/dotnet.c [hunk 1]
```diff
@@ -57,13 +57,19 @@ static ut32 clr_max_rows(int count, ...) {
 
 static int read_image_metadata_strings(RzBuffer *b, ut32 stream_start, ut32 stream_size, Pe_image_clr *clr) {
 	RzPVector /* char[] */ *strings = rz_pvector_new(free);
+	if (!strings)
```
Comment (wargio) [inline librz/bin/format/pe/dotnet.c]: from the coding prospective, please use the curly brackets also for small if/else/while/for, etc..

**Comment by wingdeans** [comment]
Just realized I also pulled in [DisaMSIL](https://ntcore.com/files/disasmsil.htm) from rizin-extras, which is licensed under
"`You can use it any context you wish. There are no license restrictions. The only thing I ask you to do is to send me your bug fixes (if any).`"

What should I do for SPDX / licensing? If it's too much hassle I'll try to rewrite.

**Comment by wargio** [comment]
> Just realized I also pulled in [DisaMSIL](https://ntcore.com/files/disasmsil.htm) from rizin-extras, which is licensed under "`You can use it any context you wish. There are no license restrictions. The only thing I ask you to do is to send me your bug fixes (if any).`"
> 
> What should I do for SPDX / licensing? If it's too much hassle I'll try to rewrite.

You need to add the Unlincense to https://github.com/rizinorg/rizin/tree/dev/LICENSES and then mark this as Unlicense

https://spdx.org/licenses/Unlicense.html

Just copy this in the `Unlincense.txt` file:
```
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or distribute this software, either in source code form or as a compiled binary, for any purpose, commercial or non-commercial, and by any means.

In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain. We make this dedication for the benefit of the public at large and to the detriment of our heirs and successors. We intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

File: librz/asm/p/asm_msil.c [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+static int disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
+	return (op->size = arch_msil_disasm (&op->buf_asm, buf, a->pc));
+}
+
+RzAsmPlugin rz_asm_plugin_msil = {
+	.name = "msil",
+	.arch = "msil",
+	.license = "PD",
```
Comment (wargio) [inline librz/asm/p/asm_msil.c]:
```suggestion
	.license = "Unlicense",
```

File: librz/asm/p/asm_msil.c [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/* radare - Public Domain - Copyright 2011-2015 - capi_x, pancake */
```
Comment (wargio) [inline librz/asm/p/asm_msil.c]: set the proper SPDX header (LGPL3) and also you can set this to your name

File: librz/asm/p/asm_msil.c [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	int o; 
+	DISASMSIL_OFFSET CodeBase = seek;
+	ILOPCODE_STRUCT ilopar[8];
+	o = DisasMSIL (buf, 16, CodeBase, ilopar, 8, &n);
+	rz_strbuf_setf (buf_asm, "%s", ilopar[0].Mnemonic);
+	return o;
+}
+
+static int disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
+	return (op->size = arch_msil_disasm (&op->buf_asm, buf, a->pc));
```
Comment (wargio) [inline librz/asm/p/asm_msil.c]: join these 2. no need for an extra jump

File: librz/asm/p/asm_msil.c [hunk 5]
```diff
@@ -1,4 +1,6 @@
-/* radare - Public Domain - Copyright 2011-2015 - capi_x, pancake */
+// SPDX-FileCopyrightText: 2011-2015 pancake <pancake@nopcode.org>
+// SPDX-FileCopyrightText: 2011-2015 capi_x
```
Comment (wargio) [inline librz/asm/p/asm_msil.c]:
```suggestion
// SPDX-FileCopyrightText: 2011-2015 capi_x
// SPDX-FileCopyrightText: 2022 wingdeans
```

File: librz/asm/arch/msil/demsil.h [hunk 6]
```diff
@@ -0,0 +1,359 @@
+// SPDX-FileCopyrightText: 2008. Daniel Pistelli.
+// SPDX-License-Identifier: Unlicense
+
+//
```
Comment (XVilka) [inline librz/asm/arch/msil/demsil.h]: Just remove these lines then

File: librz/asm/arch/msil/demsil.h [hunk 7]
```diff
@@ -0,0 +1,359 @@
+// SPDX-FileCopyrightText: 2008. Daniel Pistelli.
```
Comment (XVilka) [inline librz/asm/arch/msil/demsil.h]: Format is wrong

**Review by XVilka — CHANGES_REQUESTED** [review]
Please add tests. They are especially important in the case of a new format.

**Comment by XVilka** [comment]
@GustavoLCR please take a look also - if you have any feedback or suggestions.

**Comment by wargio** [comment]
@wingdeans i suggest for appending strings, etc.. to use RzStrBuf

**Comment by wingdeans** [comment]
Thanks. I'll keep it for now since I'm planning to attempt a rewrite of the disassembler when I'm free.
I'll definitely look at existing disassemblers for that.

File: librz/bin/format/pe/pe.c [hunk 8]
```diff
@@ -4048,6 +3876,33 @@ struct rz_bin_pe_lib_t *PE_(rz_bin_pe_get_libs)(struct PE_(rz_bin_pe_obj_t) * bi
 	return NULL;
 }
 
+RzList /* RzBinSymbol */ *PE_(rz_bin_pe_get_methods)(struct PE_(rz_bin_pe_obj_t) * bin) {
+	if (!bin || !bin->clr || !bin->clr->methoddefs) {
+		return NULL;
+	}
+	RzList /* RzBinSymbol */ *methods = rz_list_new();
```
Comment (ret2libc) [inline librz/bin/format/pe/pe.c]: should this be 
```suggestion
	RzList /* RzBinSymbol */ *methods = rz_list_newf(rz_bin_symbol_free);
```
?

File: librz/bin/format/pe/pe.c [hunk 9]
```diff
@@ -3374,8 +3201,9 @@ static int bin_pe_init(struct PE_(rz_bin_pe_obj_t) * bin) {
 
 	bin_pe_init_rich_info(bin);
 	bin_pe_init_tls(bin);
-	bin_pe_init_clr_hdr(bin);
-	bin_pe_init_metadata_hdr(bin);
+	bin_pe_init_clr(bin);
+	// bin_pe_init_metadata_hdr(bin);
+	// bin_pe_init_metadata_table(bin);
```
Comment (ret2libc) [inline librz/bin/format/pe/pe.c]:
```suggestion
```

File: librz/bin/format/pe/pe.c [hunk 10]
```diff
@@ -1213,136 +1212,23 @@ int PE_(bin_pe_get_overlay)(struct PE_(rz_bin_pe_obj_t) * bin, ut64 *size) {
...
-	if (rr < 1) {
-		goto fail;
-	}
+	Pe_image_clr *clr = RZ_NEW0(Pe_image_clr);
+	if (!clr)
+		return -1;
 
-	bin->metadata_header = metadata;
+	if (bin_pe_dotnet_init_clr(clr, bin->b, image_clr_hdr_paddr))
+		return -1;
```
Comment (ret2libc) [inline librz/bin/format/pe/pe.c]:
```suggestion
	if (bin_pe_dotnet_init_clr(clr, bin->b, image_clr_hdr_paddr)) {
		return -1;
	}

```

File: librz/bin/format/pe/dotnet.h [hunk 11]
```diff
@@ -1,85 +1,155 @@
...
-	IMAGE_DATA_DIRECTORY StrongNameSignature;
-	ULONGLONG CodeManagerTable;
-	IMAGE_DATA_DIRECTORY VTableFixups;
-	ULONGLONG ExportAddressTableJumps;
-	ULONGLONG ManagedNativeHeader;
-} CLI_HEADER, *PCLI_HEADER;
-
-#define NET_METADATA_MAGIC 0x424a5342
+#include <rz_vector.h>
+// #include <rz_util.h>
```
Comment (ret2libc) [inline librz/bin/format/pe/dotnet.h]: remove commented lines

File: librz/asm/p/asm_cil.c [hunk 12]
```diff
@@ -0,0 +1,27 @@
+// SPDX-FileCopyrightText: 2022 wingdeans
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_asm.h>
+#include "cil/cil_dis.c"
```
Comment (ret2libc) [inline librz/asm/p/asm_cil.c]: is this really necessary? can you just link the cil_dis object with asm_cil in the meson.build file?

**Comment by XVilka** [comment]
Please rebase on top of the latest `dev`.

**Comment by XVilka** [comment]
Off-by-one?
```
[XX] C:\projects\rizin\test\db\formats\pe\dotnet20 PE: corkami dotnet20.exe - open
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 's sym.Main; s
s sym..ctor; s
q!
' bins/pe/dotnet20.exe
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-0x402050
-0x40205e
+0x402051
+0x40205f
-- stderr
13
7
```

Please also fix the `clang-format`

File: librz/asm/arch/cil/cil_dis.c [hunk 13]
```diff
@@ -61,7 +54,16 @@ static int read_InlineNone(int *pos, RzAsmOp *op, const ut8 *buf, int len) {
 #define read_InlineR      read_generic_double
 
 // Special
-static int read_InlineSwitch(int *pos, RzAsmOp *op, const ut8 *buf, int len) {
+static int read_InlineMethod(int *pos, CILOp *op, const ut8 *buf, int len) {
+	if (*pos + 4 >= len) {
+		return -1;
+	}
+	rz_strbuf_appendf(&op->strbuf, " %x", rz_read_le32(buf));
+	*pos += 4;
```
Comment (wargio) [inline librz/asm/arch/cil/cil_dis.c]:
```suggestion
	*pos += sizeof(ut32);
```

**Comment by wingdeans** [comment]
That wraps up what I was planning to implement when I started. The analysis plugin in particular needs some work, but I think this is a good starting point.

File: librz/analysis/p/analysis_cil.c [hunk 14]
```diff
@@ -0,0 +1,37 @@
+// SPDX-FileCopyrightText: 2022 wingdeans
```
Comment (XVilka) [inline librz/analysis/p/analysis_cil.c]: If it's possible, please use the real name and add an email for future contact, if there will be the need.

File: librz/asm/arch/cil/cil_dis.c [hunk 15]
```diff
@@ -0,0 +1,158 @@
+// SPDX-FileCopyrightText: 2022 wingdeans
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_types.h>
+#include <rz_asm.h>
+#include <rz_endian.h>
+#include <rz_util/rz_strbuf.h>
+#include "cil_dis.h"
+
+// The read_## functions are dispatched based on the opcode `param`
```
Comment (XVilka) [inline librz/asm/arch/cil/cil_dis.c]: I recommend using the Doxygen format for comments, wherever possible, e.g. in this case it would be `\file`.

File: librz/asm/arch/cil/cil_dis.c [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+		op->byte1 = byte;
+		op->byte2 = byte = buf[pos++];
+		opcode_reader = opcode_readers_double[byte];
+	}
+
+	if (!opcode_reader.str) {
+		return -1; // Invalid
+	}
+
+	// Mneumonic
```
Comment (XVilka) [inline librz/asm/arch/cil/cil_dis.c]: Mnemonic

File: librz/bin/format/pe/dotnet.c [hunk 17]
```diff
@@ -426,6 +288,7 @@ void dotnet_parse_tilde_2(
 
 		switch (bit_check) {
 		case BIT_MODULE:
+			/*
```
Comment (XVilka) [inline librz/bin/format/pe/dotnet.c]: If it's not used - just remove this.

File: librz/bin/format/pe/dotnet.c [hunk 18]
```diff
@@ -434,1095 +297,702 @@ void dotnet_parse_tilde_2(
...
-			if (row_count > (0xFFFF >> 0x02))
-				index_size = 4;
-			else
-				index_size = 2;
+		case BIT_EVENT: {
+			ut32 index_count = clr_max_rows(3,
+				rowcounts->typedef_,
+				rowcounts->typeref,
+				rowcounts->typespec);
+			ut32 index_size = (index_count > (0xFFFF >> 0x02)) ? 4 : 2;
```
Comment (XVilka) [inline librz/bin/format/pe/dotnet.c]: It's quite common pattern, probably worth converting to macros.

File: librz/bin/format/pe/dotnet.c [hunk 19]
```diff
@@ -434,1095 +297,702 @@ void dotnet_parse_tilde_2(
...
+	}
 
-	// Set index sizes for various heaps.
-	if (tilde_header->HeapSizes & 0x01)
-		index_sizes.string = 4;
-	if (tilde_header->HeapSizes & 0x02)
-		index_sizes.guid = 4;
-	if (tilde_header->HeapSizes & 0x04)
-		index_sizes.blob = 4;
+	rr = rz_buf_fread_at(b, metadata_directory,
```
Comment (XVilka) [inline librz/bin/format/pe/dotnet.c]: Move `int rr` here

File: librz/bin/format/pe/pe.c [hunk 20]
```diff
@@ -4048,6 +3876,102 @@ struct rz_bin_pe_lib_t *PE_(rz_bin_pe_get_libs)(struct PE_(rz_bin_pe_obj_t) * bi
 	return NULL;
 }
 
+RzList /* RzBinSymbol */ *PE_(rz_bin_pe_get_clr_symbols)(struct PE_(rz_bin_pe_obj_t) * bin) {
```
Comment (XVilka) [inline librz/bin/format/pe/pe.c]: Add ownership mark

File: librz/bin/meson.build [hunk 21]
```diff
@@ -126,7 +126,7 @@ rz_bin_sources = [
   'p/bin_xbe.c',
   'p/bin_xnu_kernelcache.c',
   'p/bin_xtr_fatmach0.c',
-  'p/bin_xtr_pemixed.c',
+# 'p/bin_xtr_pemixed.c',
```
Comment (XVilka) [inline librz/bin/meson.build]: Why this change?

File: librz/bin/meson.build [hunk 22]
```diff
@@ -202,9 +202,10 @@ rz_bin_sources = [
   'format/objc/mach0_classes.c',
   'format/omf/omf.c',
   'format/p9/p9bin.c',
+  'format/pe/dotnet.c',
   'format/pe/pe.c',
   'format/pe/pe64.c',
-  'format/pe/pemixed.c',
+# 'format/pe/pemixed.c',
```
Comment (XVilka) [inline librz/bin/meson.build]: Same, why this change?

File: test/db/formats/pe/dotnet20 [hunk 23]
```diff
@@ -1,6 +1,12 @@
 NAME=PE: corkami dotnet20.exe - open
 FILE=bins/pe/dotnet20.exe
-CMDS=q!
+CMDS=<<EOF
+s sym.HelloWorld::Main; s
```
Comment (XVilka) [inline test/db/formats/pe/dotnet20]: I think, we need a bit more tests than this, especially in terms of the CIL bytecode itself, like `rz-asm` tests, also a bit more on the analysis

**Review by XVilka — CHANGES_REQUESTED** [review]
Please rebase the PR on top of the latest `dev`.
Also, ASAN jobs do not compile:
```
[790/1811] Compiling C object librz/bin/librz_bin.so.0.5.0.p/format_pe_pe.c.o
FAILED: librz/bin/librz_bin.so.0.5.0.p/format_pe_pe.c.o 
clang -Ilibrz/bin/librz_bin.so.0.5.0.p -I. -I.. -Ilibrz -I../librz -Ilibrz/include -I../librz/include -I../librz/bin/format -I../subprojects/lz4-1.9.3/lib -Isubprojects/yxml -I../subprojects/yxml -Isubprojects/sdb -I../subprojects/sdb -Isubprojects/sdb/src -I../subprojects/sdb/src -I../librz/type/parser -fcolor-diagnostics -fsanitize=address,undefined -fno-omit-frame-pointer -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -O2 -g --std=gnu99 -D_GNU_SOURCE -Werror=sizeof-pointer-memaccess -fvisibility=hidden -DASAN=1 -DRZ_ASSERT_STDOUT=1 -Wno-cpp -fPIC -DRZ_API_BIN_ONLY=1 -DRZ_PLUGIN_INCORE=1 -shared-libasan -MD -MQ librz/bin/librz_bin.so.0.5.0.p/format_pe_pe.c.o -MF librz/bin/librz_bin.so.0.5.0.p/format_pe_pe.c.o.d -o librz/bin/librz_bin.so.0.5.0.p/format_pe_pe.c.o -c ../librz/bin/format/pe/pe.c
../librz/bin/format/pe/pe.c:3958:7: error: logical not is only applied to the left hand side of this bitwise operator [-Werror,-Wlogical-not-parentheses]
                if (!methoddef->implflags & 0x01) { // not native
                    ^                     ~
../librz/bin/format/pe/pe.c:3958:7: note: add parentheses after the '!' to evaluate the bitwise operator first
                if (!methoddef->implflags & 0x01) { // not native
                    ^
                     (                          )
../librz/bin/format/pe/pe.c:3958:7: note: add parentheses around left hand side expression to silence this warning
                if (!methoddef->implflags & 0x01) { // not native
                    ^
                    (                    )
1 error generated.
[791/1811] Compiling C object librz/bin/librz_bin.so.0.5.0.p/format_pe_pe64.c.o
FAILED: librz/bin/librz_bin.so.0.5.0.p/format_pe_pe64.c.o 
clang -Ilibrz/bin/librz_bin.so.0.5.0.p -I. -I.. -Ilibrz -I../librz -Ilibrz/include -I../librz/include -I../librz/bin/format -I../subprojects/lz4-1.9.3/lib -Isubprojects/yxml -I../subprojects/yxml -Isubprojects/sdb -I../subprojects/sdb -Isubprojects/sdb/src -I../subprojects/sdb/src -I../librz/type/parser -fcolor-diagnostics -fsanitize=address,undefined -fno-omit-frame-pointer -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -O2 -g --std=gnu99 -D_GNU_SOURCE -Werror=sizeof-pointer-memaccess -fvisibility=hidden -DASAN=1 -DRZ_ASSERT_STDOUT=1 -Wno-cpp -fPIC -DRZ_API_BIN_ONLY=1 -DRZ_PLUGIN_INCORE=1 -shared-libasan -MD -MQ librz/bin/librz_bin.so.0.5.0.p/format_pe_pe64.c.o -MF librz/bin/librz_bin.so.0.5.0.p/format_pe_pe64.c.o.d -o librz/bin/librz_bin.so.0.5.0.p/format_pe_pe64.c.o -c ../librz/bin/format/pe/pe64.c
In file included from ../librz/bin/format/pe/pe64.c:5:
../librz/bin/format/pe/pe.c:3958:7: error: logical not is only applied to the left hand side of this bitwise operator [-Werror,-Wlogical-not-parentheses]
                if (!methoddef->implflags & 0x01) { // not native
                    ^                     ~
../librz/bin/format/pe/pe.c:3958:7: note: add parentheses after the '!' to evaluate the bitwise operator first
                if (!methoddef->implflags & 0x01) { // not native
                    ^
                     (                          )
../librz/bin/format/pe/pe.c:3958:7: note: add parentheses around left hand side expression to silence this warning
                if (!methoddef->implflags & 0x01) { // not native
                    ^
                    (                    )
1 error generated.
[792/1811] Compiling C object librz/bin/librz_bin.so.0.5.0.p/format_pe_dotnet.c.o
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 1.
```

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good. But could you test the CIL opcodes? 
Like it's done with the Java bytecode? https://github.com/rizinorg/rizin/blob/dev/test/db/asm/java

**Comment by XVilka** [comment]
@GustavoLCR could you please take a look, if anything is missing or wrong?

**Comment by wingdeans** [comment]
Tests are finally passing. Should I squash?

File: librz/bin/format/pe/pe.c [hunk 24]
```diff
@@ -4048,6 +3876,120 @@ struct rz_bin_pe_lib_t *PE_(rz_bin_pe_get_libs)(struct PE_(rz_bin_pe_obj_t) * bi
...
+		if (type_it) {
+			type_methods_end = ((Pe_image_metadata_typedef *)type_it->data)->methodlist;
+		}
+	}
+
+	int i = 1;
+	RzListIter *it;
+	Pe_image_metadata_methoddef *methoddef;
+	rz_list_foreach (bin->clr->methoddefs, it, methoddef) {
+		if ((type_name || type_namespace) && i >= type_methods_start) {
```
Comment (XVilka) [inline librz/bin/format/pe/pe.c]: We usually reverse the condition and call `continue` then.

File: librz/bin/format/pe/pe.c [hunk 25]
```diff
@@ -4048,6 +3876,120 @@ struct rz_bin_pe_lib_t *PE_(rz_bin_pe_get_libs)(struct PE_(rz_bin_pe_obj_t) * bi
...
+			if (i >= type_methods_end) {
```
Comment (XVilka) [inline librz/bin/format/pe/pe.c]: Can be folded into the previous condition, I guess?

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good overall. I think, we can merge it earlier and improve step by step, once you address these minor things.

*P.S. Yes, you can squash.*

File: librz/bin/format/pe/pe.c (hunk 24 above)
Comment (wingdeans) [reply librz/bin/format/pe/pe.c]: I don't think this applies here? I want the loop body to run regardless of whether the condition holds.

**Review by XVilka — APPROVED** [review]

File: librz/analysis/p/analysis_cil.c [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+		!(get_offset = analysis->binb.get_offset)) {
+		return UT64_MAX;
+	}
+
+	ut32 tok = cilop->tok;
+	switch (tok >> 24) {
+	case 0x06: { // MethodDef index
+		return get_offset(bin, 'd', tok & 0xffffff);
+	}
+	}
```
Comment (wargio) [inline librz/analysis/p/analysis_cil.c]:
```suggestion
	ut32 tok = cilop->tok;
	tok >>= 24;
	if (tok == 0x06) {
		// MethodDef index
		return get_offset(bin, 'd', tok & 0xffffff);
	}
```

File: librz/analysis/p/analysis_cil.c [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+	case CIL_OP_NOP:
+		op->type = RZ_ANALYSIS_OP_TYPE_NOP;
+		break;
+	case CIL_OP_RET:
+		op->type = RZ_ANALYSIS_OP_TYPE_RET;
+		break;
+	case 0xFE:
+		switch (cilop.byte2) {
+		}
+		break;
```
Comment (wargio) [inline librz/analysis/p/analysis_cil.c]: something is missing here, also please add default case.

File: librz/analysis/p/analysis_cil.c [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+	}
+
+	return op->size = cilop.size;
+}
+
+RzAnalysisPlugin rz_analysis_plugin_cil = {
+	.name = "cil",
+	.license = "LGPL3",
+	.op = cil_analyze_op,
```
Comment (wargio) [inline librz/analysis/p/analysis_cil.c]:
```suggestion
	.name = "cil",
	.desc = "dotnet CIL code analysis plugin",
	.license = "LGPL3",
	.arch = "cil",
	.op = cil_analyze_op,
```

File: librz/asm/arch/cil/cil_dis.c [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+int cil_dis(CILOp *op, const ut8 *buf, int len) {
+	int pos = 0;
+	if (pos >= len) { // pos + 1 > len
+		return -1;
+	}
+
+	ut8 byte;
+
+start: // Taken after a prefix opcode has been consumed
+	byte = buf[pos++];
```
Comment (wargio) [inline librz/asm/arch/cil/cil_dis.c:159]: i'm not a big fan of gotos used to create loops, but i guess is acceptable since the code is very readable.

**Review by wargio — CHANGES_REQUESTED** [review]
good job, please fix the 2 issues regarding the switch cases and then i think we can merge.

File: test/db/formats/pe/AStyleWhore [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+s sym.AStyleInterface::GetVersion; s
+s sym.AStyleInterface::OnAStyleMemAlloc; s
+s sym.AStyleInterface::OnAStyleError; s
+s sym.AStyleWhore.Properties::Resources::.ctor; s
+s sym.AStyleWhore.Properties::Resources::get_ResourceManager; s
+s sym.AStyleWhore.Properties::Resources::get_Culture; s
+s sym.AStyleWhore.Properties::Resources::set_Culture; s
+s sym.AStyleWhore::AStyleWhore::GetFilesInDir; s
+s sym.AStyleWhore::AStyleWhore::AStyleDirectory; s
+s sym.AStyleWhore::AStyleWhore::.ctor; s
```
Comment (wargio) [inline test/db/formats/pe/AStyleWhore]: just for the next time, you can do `s @ sym.xxxxxx`

File: librz/bin/p/bin_pe.inc [hunk 31]
```diff
@@ -717,3 +725,25 @@ static RzList *compute_hashes(RzBinFile *bf) {
...
+	Pe_image_clr *clr;
+
+	if (!bf || !bf->o ||
+		!(bin = bf->o->bin_obj) ||
+		!(clr = bin->clr)) {
+		return UT64_MAX;
+	}
+
+	if (type == 'd') { // MethodDef index
+		Pe_image_metadata_methoddef *methoddef = rz_list_get_n(clr->methoddefs, idx - 1);
```
Comment (wargio) [inline librz/bin/p/bin_pe.inc]: another note: using rz_list in this way is kinda slow, because it will loop till `n` equals `idx - 1`, so this is like a big bottleneck.
I had this issue on dex files, which increased the loading time by 1000%

**Review by wargio — APPROVED** [review]
I'm ok with this, but the bottleneck can be heavy, unless is rarely called.

File: librz/bin/p/bin_pe.inc (hunk 31 above)
Comment (wingdeans) [reply librz/bin/p/bin_pe.inc]: Makes sense. I'll switch to PVectors
